crates/ruma-signatures/src/keys.rs

```diff
@@ -42,27 +90,59 @@ impl Ed25519KeyPair {
     ///
     /// Returns an error if the public and private keys provided are invalid for the implementing
     /// algorithm.
-    pub fn new(document: &[u8], version: String) -> Result<Self, Error> {
-        let keypair = RingEd25519KeyPair::from_pkcs8(document)
-            .map_err(|error| Error::new(error.to_string()))?;
+    ///
+    /// Returns an error when the PKCS#8 document had a public key, but it doesn't match the one
+    /// generated from the private key. This is a fallback and extra validation against
+    /// corruption or
+    pub fn from_der(document: &[u8], version: String) -> Result<Self, Error> {
+        let oak = OneAsymmetricKey::from_der(document).map_err(|e| Error::new(format!("{}", e)))?;
 
-        Ok(Self { keypair, version })
+        Self::from_pkcs8_oak(oak, version)
+    }
+
+    /// Constructs a key pair from [`pkcs8::OneAsymmetricKey`].
+    pub fn from_pkcs8_oak(oak: OneAsymmetricKey, version: String) -> Result<Self, Error> {
+        Self::new(oak.algorithm.oid, oak.private_key, oak.public_key, version)
+    }
+
+    /// Constructs a key pair from [`pkcs8::PrivateKeyInfo`].
+    pub fn from_pkcs8_pki(oak: PrivateKeyInfo, version: String) -> Result<Self, Error> {
+        Self::new(oak.algorithm.oid, oak.private_key, None, version)
+    }
+
+    /// This is due to `ring` messing up the formatting of the PKCS#8 document,
+    /// corrected here for backwards compatibility.
+    /// See: https://github.com/briansmith/ring/issues/1299
+    fn correct_faulty_ring_private_key(key: &[u8]) -> &[u8] {
+        if key.len() == 34 && &key[..2] == &[0x04, 0x20] {
+            return &key[2..];
+        } else {
+            return key;
+        }
     }
```

Comment:
Because we used `ring` in the past, we have technical debt for anyone who might want to use the library to parse `ring`'s documents, this quickly fixes it, i think this is reasonable enough to keep around for the moment.

Reply:
Sure. But the `return <expr>;`s can just be `<expr>`s 😉